File: docs/test-cases/injection.java
// Test case: injection (A05:2025)
package com.example.app;

import java.sql.*;
import javax.servlet.http.*;
import org.springframework.web.bind.annotation.*;
import freemarker.template.*;

@RestController
public class UserController {

    @GetMapping("/users")
    public String getUser(HttpServletRequest req) throws Exception {
        String userId = req.getParameter("id");
        Connection conn = DriverManager.getConnection("jdbc:postgresql://db/app");
        Statement stmt = conn.createStatement();
        // BUG: SQL injection via string concatenation
        ResultSet rs = stmt.executeQuery("SELECT * FROM users WHERE id = " + userId);
        return rs.next() ? rs.getString("name") : "none";
    }

    @GetMapping("/convert")
    public String convertFile(@RequestParam String filename) throws Exception {
        // BUG: shell injection — Runtime.exec with user input parsed by the shell
        Process p = Runtime.getRuntime().exec(new String[]{"sh", "-c", "convert " + filename + " out.png"});
        return "ok: " + p.waitFor();
    }

    @GetMapping("/greet")
    public String greet(@RequestParam String name) throws Exception {
        Configuration cfg = new Configuration(Configuration.VERSION_2_3_31);
        // BUG: server-side template injection — user input compiled as a FreeMarker template
        Template t = new Template("greet", new java.io.StringReader("Hello " + name + "!"), cfg);
        java.io.StringWriter out = new java.io.StringWriter();
        t.process(new java.util.HashMap<>(), out);
        return out.toString();
    }
}

File: docs/test-cases/injection.go
// Test case: injection (A05:2025)
package main

import (
	"database/sql"
	"fmt"
	"net/http"
	"os/exec"
	"text/template"

	_ "github.com/lib/pq"
)

var db *sql.DB

func getUser(w http.ResponseWriter, r *http.Request) {
	userID := r.URL.Query().Get("id")
	// BUG: SQL injection via fmt.Sprintf into db.Query
	rows, err := db.Query(fmt.Sprintf("SELECT name FROM users WHERE id = %s", userID))
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer rows.Close()
}

func convertFile(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("filename")
	// BUG: shell injection — user input passed to sh -c
	out, _ := exec.Command("sh", "-c", "convert "+filename+" out.png").CombinedOutput()
	w.Write(out)
}

func greet(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	// BUG: server-side template injection — user input compiled into a template
	t, _ := template.New("greet").Parse(fmt.Sprintf("Hello %s!", name))
	t.Execute(w, nil)
}

func main() {
	http.HandleFunc("/users", getUser)
	http.HandleFunc("/convert", convertFile)
	http.HandleFunc("/greet", greet)
	http.ListenAndServe(":8080", nil)
}
